Pack oemof.solph results into a single MultiIndex Dataframe

Bachibouzouk opened the pull request:
## Summary of the discussion

Introduce the MultiIndex Dataframe we discussed about in #10 

## Type of change (CHANGELOG.md)

### Added
- Add functions to create a multi-index dataframe out of the raw oemof results

Bachibouzouk committed:
Add functions to create a single Dataframe containing oemof results

Those are in a separate module as they could very well be implemented
into oemof.solph

Bachibouzouk committed:
Add the oemof results Dataframe to an OTP version of Calculator

Bachibouzouk committed:
Populate the MultiIndex DataFrame columns and start describing some calculations

ciaradunks commented on examples/scripts/compute.py:
```diff
@@ -84,7 +85,8 @@
     # extract parameters and results
     params = parameter_as_dict(es)
     results = m.results()
```
can this be deleted now because the results parameter is not used?

Bachibouzouk replied:
Yes after the PR is finalized we can do cleanup

ciaradunks reviewed (commented):
This is going to be a great new feature! I really like the way it is structured and being able to access/calculate additional results more clearly.

1. Most of the examples do not work because of a conflict in energy carriers, one of us should resolve this - probably in this PR?
2. It would be good to talk about the next steps, what needs doing and who should do what

ciaradunks commented:
### Calculation example (upfront investment costs)

#### Parameters we need:

- From inputs: CAPEX
- From raw results: optimized capacity

Upfront investment cost = capex input * optimized capacity
total upfront investment costs = sum of upfront investment cost for each component

Bachibouzouk commented on examples/scripts/compute.py:
```diff
@@ -25,7 +26,7 @@
 # -------------- USER INPUTS --------------
 # list of scenarios to be evaluated
-scenarios = ["general_add_cost_inputs"]
+scenarios = ["wefe_pv_panel"]
```
this is just for testing purposes

Bachibouzouk commented on src/oemof_tabular_plugins/datapackage/post_processing.py:
```diff
@@ -0,0 +1,339 @@
+
+RAW_OUTPUTS = ["investments"]
+PROCESSED_RAW_OUTPUTS = ["flow_min", "flow_max", "aggregated_flow"]
+RAW_INPUTS = [
```
this can be expanded @ciaradunks

Bachibouzouk commented on src/oemof_tabular_plugins/datapackage/post_processing.py:
```diff
@@ -0,0 +1,339 @@
+    )
+
+
+CALCULATED_OUTPUTS = [
```
@ciaradunks - This is where one defines the calculations, note that if you want to add yours, you also have to define the function which is used under "operation"

Bachibouzouk commented on src/oemof_tabular_plugins/general/post_processing/post_processing.py:
```diff
@@ -650,18 +656,29 @@ def create_costs_table(all_scalars, results, capacities_df, storage_capacities_d
-    calculator = Calculator(params, results)
-
+    calculator = OTPCalculator(params, es, dp_path)
+    print(calculator.df_results)
```
@ciaradunks  this is just for testing purposes, you can set a breakpoint here to have a look at the multi-index dataframe

Bachibouzouk reviewed (commented):
@ciaradunks 

Now you can test if you can run it and have a look at the final DataFrame, I will then rework the commits for a final review before we merge

ciaradunks commented:
@Bachibouzouk I have tested it and the dataframe looks great! And variable costs have been calculated/implemented in the dataframe :)
1. Do you want me to add to the input/output parameters in this PR? and additional calculations? or we can do this after the merge?

Bachibouzouk commented:
> @Bachibouzouk I have tested it and the dataframe looks great! And variable costs have been calculated/implemented in the dataframe :)
> 
>     1. Do you want me to add to the input/output parameters in this PR? and additional calculations? or we can do this after the merge?

Yes please :) Add them so that we merge at once

Bachibouzouk committed:
Add validation of calculations and rearrange code a bit

Bachibouzouk committed:
Pseudo code for package validation for specific components

Bachibouzouk committed:
Provide a structure for post-processing the raw inputs of specific facades

Adapt the MIMO Facade package validation method

Bachibouzouk commented on src/oemof_tabular_plugins/datapackage/post_processing.py:
```diff
@@ -119,6 +119,9 @@ def infer_busses_carrier(energy_system):
                 if hasattr(node, attribute):
 
                     bus_label = getattr(node, attribute).label
+                    print(
```
on the line 118 where you have  `for attribute in ("bus", "from_bus"):`

you can simply add the column name you want to read for inferring carrier `for attribute in ("bus", "from_bus", "from_bus_0", "from_bus_1"):`

ciaradunks replied:
it's actually more complicated than this because the MIMO now there are problems with having one carrier but multiple busses associated with it. Should I write a different infer_busses_carrier method for MIMO?

![image](https://github.com/rl-institut/oemof-tabular-plugins/assets/53213814/6afa6eb5-a8d8-4664-b3d1-f3da74e45c90)

Bachibouzouk replied:
It is the reverse, the error says one bus has two carriers, which does not make sense indeed

ciaradunks replied:
yes true, but because of the MIMO only having one carrier and two busses (so 'water' being the carrier as this is the primary), this leads to this error with the ac-elec-bus having these two carriers attached to it

Commit:
Raw inputs included in df_results and fix for carrier/multiple bus/MIMO issue

paulapreuss committed:
add variable, opex and annuity costs to results dataframe

paulapreuss committed:
add new functions to get sub-tables from results dataframe (WIP)

paulapreuss committed:
expand raw inputs list

capacity potential should be included in the capacities sub-table, so we already add it to the "mother" dataframe to be extracted later

ciaradunks committed:
Update results dataframe calculations

ciaradunks committed:
Adapted scenarios to work with new results formatting

ciaradunks committed:
Modified gitignore to include all scenarios

ciaradunks committed:
set numpy version==1.26.0 in pyproject.toml

ciaradunks committed:
Added KPI results table and saved as CSV

ciaradunks committed:
Deleted old post-processing functions

ciaradunks committed:
Deleted print statements from MIMO converter

ciaradunks committed:
Quick fix for correcting cost results and including renewable share

ciaradunks committed:
Add oemof-tabular dev and oemof-industry saltwater requirement

Bachibouzouk commented on examples/scripts/compute.py:
```diff
@@ -25,18 +26,25 @@
 # -------------- USER INPUTS --------------
 # list of scenarios to be evaluated
-scenarios = ["general_add_cost_inputs"]
+scenarios = ["cp_nigeria_usungwe_example"]
```
I guess it is was overseen @paulapreuss ?

paulapreuss replied:
Yes, I'd say this was an oversight leftover from testing.

Bachibouzouk commented on src/oemof_tabular_plugins/datapackage/post_processing.py:
```diff
@@ -0,0 +1,828 @@
+
+def compute_opex_fix_costs(results_df):
+    """Calculates yearly opex costs by multiplying opex with optimized capacity (investments)"""
+    if "opex_fix" not in results_df.index:
```
normally this check is not needed as "opex_fix" is listed as a function argument

Bachibouzouk committed:
Code cosmetics
